### ISSUE TITLE
decimal separator in math mode for spanish

### ISSUE THREAD

**Issue by asaltanubes** [typst/typst#1093]
### Description

In Spanish the decimal separator is a comma instead of a period, in latex when using Babel in Spanish mode when in math mode the decimal separator is changed automatically. It would be nice to have the same treatment when using text(lang: "en")

### Use Case

#set text(lang: "es")
$3.1415$
should output something like 
![image](https://user-images.githubusercontent.com/56401901/235989705-c196c021-92a1-4e72-ae05-629b457d7ae7.png)
instead of 
![image](https://user-images.githubusercontent.com/56401901/235990025-60118022-1224-42c2-b44f-119d878e1f13.png)

**Comment by owiecc** [comment]
This is not limited to Spanish language: https://en.wikipedia.org/wiki/Decimal_separator#Conventions_worldwide

**Comment by asaltanubes** [comment]
Found a solution, it isnt the cleanest but it works.
`
#show regex("\d+\.\d+"): it => {show ".": ","
  it}
`
if only wanted in math mode
`
#show math.equation: it => {
show regex("\d+\.\d+"): it => {show ".": ","
  it}
it
}
`

**Comment by binarynoise** [comment]
The math-mode rule seems broken, it requires semicolons or line breaks before the `it`s now

**Comment by asaltanubes** [comment]
True single quote code block from github deleted the newlines. I fixed it.

**Comment by DegrangeM** [comment]
Thanks for the code, even though it seems important to be avaible natively in typst (french here).

**Comment by FeldrinH** [comment]
https://github.com/typst/typst/issues/448 is more or less the same issue, but without a concrete proposal.

**Comment by Blezz-tech** [comment]
I hope this feature will be added in the near future

**Comment by remiangot** [comment]
It should also handle spaces and separator for thousands like https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString

30000.65 → "30,000.65" where English is the default language, or
30000.65 → "30.000,65" where German is the default language, or
30000.65 → "30 000,65" where French is the default language

**Comment by mkpoli** [comment]
An alternative solution using a function
```typst
#let n(num, decimal: ".", thousands: ",") = {
  let parts = str(num).split(".")
  let decimal_part = if parts.len() == 2 { parts.at(1) }
  let integer_part = parts.at(0).rev().clusters().enumerate()
    .map((item) => {
      let (index, value) = item
      return value + if calc.rem(index, 3) == 0 and index != 0 {
        thousands
      }
    }).rev().join("")
  return integer_part + if decimal_part != none { decimal + decimal_part }
}

#let d = n.with(decimal: ".", thousands: ",") // English Style
#let c = n.with(decimal: ",", thousands: ".") // Continental Style
#let s = n.with(decimal: ",", thousands: " ") // SI Style (English)
#let f = n.with(decimal: ",", thousands: " ") // SI Style (French)

#for x in (1.2, 65536, 3800.25) {
  [
    *#x* \
    #raw("d" + "(" + str(x) + ")") = #d(x) \
    #raw("c" + "(" + str(x) + ")") = #c(x) \
    #raw("f" + "(" + str(x) + ")") = #s(x) \
    #raw("s" + "(" + str(x) + ")") = #f(x) \
    \
  ]
}

```

Resulting

![image](https://github.com/typst/typst/assets/3502597/c169e486-2a93-4c1e-a2f5-0f53811deb3c)

Due to #229, as far as I know, you cannot replace both decimal point and thousands separator at the same time using `show` rules and `regex()`.

**Comment by asaltanubes** [comment]
I've found a way to use the previous solution with a show rule. It is very hacky and will definitely break somewere. Right now it breaks in headings because typst treats it as text and the solution i used before of checking if it.func() was heading no longer works, the only way i found to prevent this is by showing the headings manually after the show of numbers.

```typst
// This helper function converts content to strings (issue 2196 https://github.com/typst/typst/issues/2196)
#let to-string(content) = {
  if content.has("text") {
    content.text
  } else if content.has("children") {
    content.children.map(to-string).join("")
  } else if content.has("body") {
    to-string(content.body)
  } else if content == [ ] {
    " "
  }
}

#let n(num, decimal: ".", thousands: ",") = {
  let parts = str(num).split(".")
  let decimal_part = if parts.len() == 2 { parts.at(1) }
  let integer_part = parts.at(0).rev().clusters().enumerate()
    .map((item) => {
      let (index, value) = item
      return value + if calc.rem(index, 3) == 0 and index != 0 {
        thousands
      }
    }).rev().join("")
  return integer_part + if decimal_part != none { decimal + decimal_part }
}

// This regex checks for numbers with decimal part or that have 4 or more non decimal
// digits without thousand separator (1000,265.64 wont be fixed to 1.000.265,64)
#show regex("^\d{4,}(\.\d+)?|\d+\.\d+$"): it => {
  if it.func() != heading {n(to-string(it), decimal: ",", thousands: ".")}
  else {it}
}

// make headings right
#show heading: it => {
    set par(first-line-indent: 0pt)
    if it.outlined{
      numbering("1.1", ..counter(heading).at(it.location())) + " "
    }
    it.body
    linebreak()
    v(10pt)
  }
```

**Comment by Andrew15-5** [comment]
@asaltanubes, looks like there are a few other options:

```typ
#show math.equation: it => {
  // show regex("\d+\.\d+"): num => {
  //   show ".": math.class("normal", ",")
  //   num
  // }
  show regex("\d+\.\d+"): num => num.text.replace(".", ",")
  it
}

$6541.5654$
```

![image](https://github.com/user-attachments/assets/df120da3-d5cd-4e7e-afce-42c3f58580ff)

---

I can say that this is also a problem in Russian. Moreover, I feel like as a user I would like to be able to write `$0,5$` in Typst, and it would print 0,5 without the space. See https://github.com/typst/typst/issues/5272.

**Comment by laurmaedje** [comment]
Automatic locale-aware number formatting will happen, but `$0,5$` will most likely not happen because the comma already has meaning in math argument lists.

**Comment by Andrew15-5** [comment]
> `$0,5$` will most likely not happen because the comma already has meaning in math argument lists.

Yeah, I thought of that. The only solution I see is to treat them differently with the use of a space. But since there are probably instances where all numbers are mashed together, this will break such cases. Perhaps, it could have the "mode" where it will hint user of writing `1,5,9` — "in `text.lang` you should/must use space to separate numbers". So it will give an error, or fallback to using "number list syntax" and give a warning.

So, I think it's possible, but there will be some small downsides.

**Comment by ilmaisin** [comment]
What about using semicolon as a list separator in the source, and parsing both point and comma as decimal separators?

**Comment by ilmaisin** [comment]
https://github.com/Mc-Zen/zero

There is a package called zero that can do it. Not to say that this shouldn't be implemented in typst itself, but as information for end users who came here looking for how to do it.

**Comment by DegrangeM** [comment]
Zero package is nice, but it require to apply a function to each number which is annoying. (and also I am still hoping that typst add soon natively local aware number)

**Comment by goyalyashpal** [comment]
also the indic separators which go as: 3 initially (ones, tens, hundreds), then in groups of 2s (one, ten),  ...  i will be calling it as (3+2*), and the english one is (3*)

example:

```
    98,76,54,321.01234567

    98 crores
       76 lakhs
          54 thousands
             3 hundred 21
```

<strike>
technically, in accounts / economy matters, this pattern of (2*) continues only for 3 groups i.e. upto crores; then the (3+2*) repeats again.

so:

```
12,390,00,00,000.00
     ^~~~ 12 thousand, 390 crores
```

</strike>

[W5AMyLXFp5k]:
https://youtu.be/W5AMyLXFp5k " ₹70,000 Crore Scam In Bihar? NO, Just Another INDI Alliance Bluff "



_edit: this economy thing is not written in full figures;  rather it starts with crores itself then i.e. `12,390.0 crores` not `12,390,00,00,000`.  so yeah, the normal (3+2*) rule applies._

**Comment by AndyP3r3z** [comment]
I have something to add: this is not actually language dependent, but **country dependent**.  In other Spanish-speaking countries we do use the comma as a thousand separator, and the dot as decimal separator. So, setting `lang: "es"` in the whole document would format the numbers in an undesired way for other countries that speak Spanish but are not Spain...

**Comment by Andrew15-5** [comment]
The title says ["locale"](https://en.wikipedia.org/wiki/Locale_(computer_software)), which normally implies at least [language](https://typst.app/docs/reference/text/text/#parameters-lang) and [region](https://typst.app/docs/reference/text/text/#parameters-region)/country. This is the most common format, but region can be omitted.

**Comment by Hannes-Steffenhagen-bd** [comment]
Yes, language is not really sufficient here. There are plenty of cases where things like number formats different between different regions/cultures speaking the same language. And it will probably need to be somewhat adjustable to cover corner cases (for instance – what if I want to translate some german official text to english? Should I use `.` as the decimal separator and `,` for thousands, as in the UK and most other english speaking countries? Or should I keep the number formatting consistent between translations? I think there are arguments for either).

**Comment by bugsbugsbux** [comment]
rather than language/locale/country dependant this should be a style setting, so you can set and change it however and whenever you want regardless of other settings.

**Comment by DegrangeM** [comment]
It should have both : option to configure everything like we want, but with default value that depend on the language/country selected.

**Comment by stanlrt** [comment]
[https://typst.app/universe/package/zero/](https://typst.app/universe/package/zero/)

**Comment by DegrangeM** [comment]
We need something directly in typst, not a package, and working immediatly, not requiring applying a function to the number.
